### PR TITLE
Low Carbon Hub, installation and get readings - part 1

### PR DESCRIPTION
**Pull request by jamesjefferies**

### CODE REVIEW

File: app/models/meter.rb [hunk 1]
```diff
@@ -33,20 +38,32 @@ class Meter < ApplicationRecord
   scope :inactive, -> { where(active: false) }
   scope :no_amr_validated_readings, -> { left_outer_joins(:amr_validated_readings).where(amr_validated_readings: { meter_id: nil }) }
 
-  enum meter_type: [:electricity, :gas]
+
+  CREATABLE_METERS = [:electricity, :gas].freeze
+  PSEUDO_METERS = [:solar_pv, :exported_solar_pv, :electricity].freeze
+  METER_TYPES = (CREATABLE_METERS + PSEUDO_METERS).uniq
+
+  enum meter_type: METER_TYPES
```
Comment (jamesjefferies) [inline app/models/meter.rb]: I'm not 100% keen on this, but it does the trick - we have normal electricity meters and pseudo electricity meters.

File: app/models/meter.rb [hunk 2]
```diff
@@ -33,20 +38,32 @@ class Meter < ApplicationRecord
...
-  validates_format_of :mpan_mprn, with: /\A[1-3]\d{12}\Z/, if: :electricity?, message: 'for electricity meters should be a 13 digit number'
+  validates_format_of :mpan_mprn, with: /\A[1-3]\d{12}\Z/, if: :traditional_mpan?, message: 'for electricity meters should be a 13 digit number'
+  validates_format_of :mpan_mprn, with: /\A[6,7,9]\d{13}\Z/, if: :pseudo_mpan?, message: 'for electricity meters should be a 13 digit number'
   validates_format_of :mpan_mprn, with: /\A\d{1,10}\Z/, if: :gas?, message: 'for gas meters should be a 1-10 digit number'
 
   def fuel_type
     meter_type.to_sym
   end
 
+  def self.non_gas_meter_types
```
Comment (jamesjefferies) [inline app/models/meter.rb]: Previously we validated and aggregated gas or electricity, we now need to aggregate all of the non-gas meter types

File: lib/tasks/deployment/20190918115440_set_up_long_furlong.rake [hunk 3]
```diff
@@ -0,0 +1,16 @@
+namespace :after_party do
+  desc 'Deployment task: set_up_long_furlong'
+  task set_up_long_furlong: :environment do
```
Comment (jamesalmond) [inline lib/tasks/deployment/20190918115440_set_up_long_furlong.rake]: Not sure about putting live school data in here. Also means we can't do the setup in the admin interface on test?

File: spec/services/amr/low_carbon_hub_installation_factory_spec.rb [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    it 'creates the meters and initial readings' do
+      expect(low_carbon_hub_api).to receive(:full_installation_information).with(rbee_meter_id).and_return(information)
+      expect(low_carbon_hub_api).to receive(:first_meter_reading_date).with(rbee_meter_id).and_return(start_date)
+      expect(low_carbon_hub_api).to receive(:download).with(rbee_meter_id, school.urn, start_date, end_date).and_return(readings)
+
+      expect(amr_data_feed_config).to_not be nil
+
+      factory = LowCarbonHubInstallationFactory.new(school: school, rbee_meter_id: rbee_meter_id, low_carbon_hub_api: low_carbon_hub_api, amr_data_feed_config: amr_data_feed_config)
+      expect { factory.perform }.to change { Meter.count }.by(3)
```
Comment (jamesalmond) [inline spec/services/amr/low_carbon_hub_installation_factory_spec.rb]: I'd maybe scope all these meter checks to the school. i.e `school.meters.count`. Might make it clearer?

File: app/models/low_carbon_hub_installation.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
+# == Schema Information
+#
+# Table name: low_carbon_hub_installations
+#
+#  created_at    :datetime         not null
+#  id            :bigint(8)        not null, primary key
+#  information   :json
+#  rbee_meter_id :integer
```
[the remark this answers is not in the document]
Comment (jamesjefferies) [reply app/models/low_carbon_hub_installation.rb]: Fair enough - let's store this as a string....

File: app/models/meter.rb (hunk 1 above)
Comment (jamesjefferies) [reply app/models/meter.rb]: CREATABLE_METER_TYPES = [🔌 , 😲 ]

METER_TYPES = explicit

File: app/models/meter.rb [hunk 6]
```diff
@@ -2,51 +2,70 @@
...
+    meter_types = Meter::METER_TYPES.dup
+    meter_types.delete(:gas)
+    meter_types
```
[the remark this answers is not in the document]
Comment (jamesjefferies) [reply app/models/meter.rb]: Yepp

File: app/services/amr/low_carbon_hub_installation_factory.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    end
+
+    def perform
+      installation = LowCarbonHubInstallation.where(school_id: @school.id, rbee_meter_id: @rbee_meter_id,).first_or_create!
+      installation.update(information: information.to_json)
+
+      LowCarbonHubDownloadAndUpsert.new(
+        low_carbon_hub_installation: installation,
+        start_date: first_reading_date,
+        end_date: first_reading_date + 1.day,
```
[the remark this answers is not in the document]
Comment (jamesjefferies) [reply app/services/amr/low_carbon_hub_installation_factory.rb]: I'll add a comment, but it is literally just to get some data back to create the meters

File: lib/tasks/deployment/20190918115440_set_up_long_furlong.rake (hunk 3 above)
Comment (jamesjefferies) [reply lib/tasks/deployment/20190918115440_set_up_long_furlong.rake]: We can leave it out for now, the UI will be along imminently

File: spec/services/amr/low_carbon_hub_installation_factory_spec.rb (hunk 4 above)
Comment (jamesjefferies) [reply spec/services/amr/low_carbon_hub_installation_factory_spec.rb]: Yes, that's a fair shout

**Pull request merged**